src/memo/memo.ml

```diff
@@ -0,0 +1,335 @@
+      try
+        if Dag.is_child rev_dep dep_node |> not then
+          Dag.add global_dep_dag rev_dep dep_node
+      with Dag.Cycle cycle ->
+        Cycle_error.E {
+          stack = stack;
+          cycle = List.map cycle ~f:(fun node -> node.Dag.data)
+        } |> raise
+
+  let get_deps t inp =
+    match Table.find t.cache inp with
+    | None | Some { state = Running _; _ } -> None
+    | Some { state = Done cv; _ } ->
+      Some (List.map cv.deps ~f:(fun (Last_dep.T (n,_u)) ->
+        (Function_name.to_string n.spec.name, ser_input n)))
+
+  let create name ?(allow_cutoff=true) output f =
+    let name = Function_name.make name in
+    { cache = Table.create 1024
+    ; spec = { name; input = (module Input); output; allow_cutoff; witness = W }
+    ; f
+    }
+
+  let compute t inp ivar dep_node =
+    (* define the function to update / double check intermediate result *)
+    (* set context of computation then run it *)
+    push_stack_frame (T dep_node) (t.f inp) >>= fun res ->
+    (* update the output cache with the correct value *)
+    let deps =
+      Dag.children (dag_node dep_node)
+      |> List.map ~f:(fun { Dag.data = Dep_node.T node; _ } ->
+        match node.state with
+        | Running _ -> assert false
+        | Done res ->
+          Last_dep.T (node, res.data)
+      )
+    in
+    dep_node.state <- Done (Cached_value.create res ~deps);
+    (* fill the ivar for any waiting threads *)
+    Fiber.Ivar.fill ivar res >>= fun () ->
+    Fiber.return res
+
+  (* the computation that force computes the fiber *)
+  let recompute t inp (dep_node : _ Dep_node.t) =
+    (* create an ivar so other threads can wait for the computation to
+       finish *)
+    let ivar : 'b Fiber.Ivar.t = Fiber.Ivar.create () in
+    dep_node.state <- Running (Run.current (), ivar);
+    compute t inp ivar dep_node
+
+  let exec t inp =
+    match Table.find t.cache inp with
+    | None ->
+      let ivar = Fiber.Ivar.create () in
+      let dep_node : _ Dep_node.t =
+        { id = Id.gen ()
+        ; input = inp
+        ; spec = t.spec
+        ; dag_node = lazy (assert false)
+        ; state = Running (Run.current (), ivar)
+        }
+      in
+      let dag_node : Dag.node =
+        { info = Dag.create_node_info global_dep_dag
+        ; data = Dep_node.T dep_node
+        }
+      in
+      dep_node.dag_node <- lazy dag_node;
+      Table.add t.cache inp dep_node;
+      add_rev_dep dag_node >>= fun () ->
+      compute t inp ivar dep_node
+    | Some dep_node ->
+      add_rev_dep (dag_node dep_node) >>= fun () ->
+      match dep_node.state with
+      | Running (run, fut) ->
+        if Run.is_current run then
+          Fiber.Ivar.read fut
+        else
+          recompute t inp dep_node
+      | Done cv ->
+        Cached_value.get cv >>= function
+        | Some v -> Fiber.return v
+        | None -> recompute t inp dep_node
+
+  let peek t inp =
+    (* XXX: this doesn't add a reverse dependency, which is wrong *)
```

Comment:
Should this be addressed? I'd rather replace this with a pointer to a bug or something a bit actionable. Or just remove the XXX.

Reply:
Indeed, I opened a ticket and added a reference